File: trials/storage.py
"""

This is a class that has three functions.
1. initialises an object
2. add an element(no duplicate elements allowed)
3. remove an element
4. getRandom() : get a random number from the object already inserted(with equal probability of picking for all elements)

"""

import random

class Store:
    def __init__(self):
        self.store = []
        self.index_map = {}
        
    def add(self, element):
        if element not in self.index_map:
            self.store.append(element) # add the element to the store
            self.index_map[element] = len(self.store) - 1 # store the index of the element in the hash map for quick access
        
    def remove(self, element):
        if element in self.index_map:
            index = self.index_map[element]
            last_element = self.store[-1] # get the last element in the store
            
            # Move the last element to the index of the element to be removed
            self.store[index] = last_element
            
            # Update the index of the last element only if it's different from the element being removed
            if last_element != element:
                self.index_map[last_element] = index
            
            self.store.pop() # remove the last element from the store
            del self.index_map[element] # remove the element from the index map
            
    def getRandom(self):
        if not self.store:
            return None
        return random.choice(self.store)  # return a random element from the store
    

def main():
    store = Store()
    store.add(1)
    store.add(2)
    store.add(3)
    print(store.getRandom())  # Should print a random element from 1, 2, or 3
    store.remove(1)
    print(store.getRandom())  # Should print a random element from 1 or 3
    
    
if __name__ == "__main__":
    main()

File: trials/src/storage.cpp
#include<bits/stdc++.h>
using namespace std;

class Store{
    unordered_map<int, int> elementIndexMap;
    vector<int> elements;
    public:
        int add(int element){
            if(elementIndexMap.find(element) != elementIndexMap.end()){
                return 0; // element already exists
            }
            elements.push_back(element);
            elementIndexMap[element] = elements.size() - 1;
            return 1; // element added successfully
        }

        int remove(int element){
            if(elementIndexMap.find(element) == elementIndexMap.end()){
                return 0; // element does not exist
            }
            int index = elementIndexMap[element];
            int lastElement = elements.back(); // get the last element
            // Move the last element to the position of the element to be removed
            elements[index] = lastElement; // replace the element to be removed with the last element
            // Update the index of the last element in the map
            elementIndexMap[lastElement] = index; // update the index of the last element
            // Remove the last element from the vector and map
            elements.pop_back(); // remove the last element
            elementIndexMap.erase(element); // remove the element from the map
            // No need to update the index of the last element in the map since it was already
            // moved to the position of the removed element
            return 1; // element removed successfully
        }

        int getRandom(){
            if(elements.empty()){
                return -1; // no elements to return
            }
            int randomIndex = rand() % elements.size(); // generate a random index
            return elements[randomIndex]; // return the element at the random index
        }

        //function to display the elements (for testing purposes)
        void display(){
            for(int element : elements){
                cout << element << " ";
            }
            cout << endl;
        }

        void clear(){
            elements.clear(); // clear the vector
            elementIndexMap.clear(); // clear the map
        }
};

int main() {
    Store store;
    store.add(1);
    store.add(2);
    store.add(3);
    store.add(2); // should not add duplicate
    store.add(4);
    store.display(); // should print 1 2 3
    cout << "Random Element: " << store.getRandom() << endl; // should print a random element from 1, 2, or 3
    store.remove(4);
    store.display(); // should print 1 2 3
    cout << "Random Element after removing 4: " << store.getRandom() << endl; // should print a random element from 2 or 3
    return 0;
}